### ISSUE TITLE
SSO support

### ISSUE THREAD

**Issue by gbaman** [gbaman/NI-Jam-Information-System#38]
It's been on the todo list for quite a while now, but ideally NIJIS should move over to a single sign on provider eventually. The alternative is implementing MFA ourselves, but in the end, piggybacking off a large SSO provider is just going to make the most sense, instead of doing everything ourselves.
This brings with it major security benefits (alongside MFA support).

I am not against maintaining the standalone login page as an option in NIJIS (for other orgs?), but ideally this should be phased out for the NI Jam.

Of the 2 main providers we already use, Microsoft and Google, I would lean towards Microsoft as have experience already with their SSO integration for Flask on other projects.

# How could it be implemented on an existing installation?
I am thinking the following process.
1. Use MSAL library to perform the verification on the user login.
2. Once it returns back to the app with the "id_token" (assuming successful), we can open that up to get the email address.
3. With that email address, we can then log the user in with the system (and associate that login user with that session) by matching to the email address.
4. If the email address doesn't exist, we can simply create a new row in `login_users` for that user. 

All this will depend on groups to be properly set up on the Microsoft Entra ID side, so it's only users that have the correct group permission, can log into NIJIS. This could also mean migrating over to using groups on Entra ID side as opposed to the built in group system in NIJIS.

Because the current login_users table stays intact, all other code should continue to just work. It also allows if needed, both login options to work side by side at least initially.

The only real disadvantage (bar the work to build it) with this is all users will need email addresses changed and organisation email addresses (even if they aren't used) set up. It would be possible though to maintain a second "personal email address" field within the `login_users` table if needed, but things like password resets etc that would handled on the Microsoft side, not ours then.

Any thoughts or suggestions?

**Comment by vlee489** [comment]
I agree with the move over to using SSO to handle login, however I think an argument can be made to use something more generic than the MSAL library and opt for something like OpenId Connect which builds off oauth2 and can be used more generally with other auth providers such Google workspace, Authentik etc.

**Comment by gbaman** [comment]
@vlee489 do you have a preferred library for Python for OpenId Connect? 
MSAL is a nice drop in for Flask. I'm all for exploring other options, although in the end, all we really need to be considering is what works for the NI Jam, as opposed to having the need to make it generic for other orgs as I'm not aware of others using it anyway. Adding a 3rd party element in the middle does I guess also add another link in the chain that could go wrong, but very much not against it if it isn't a whole lot more work.
